From d0a6af73ed81ca99214e05fcad7b0a045c7ab991 Mon Sep 17 00:00:00 2001
From: Jonas Karlman <jonas@kwiboo.se>
Date: Sun, 26 Apr 2026 23:32:03 +0000
Subject: [PATCH 167/280] WIP: drm/bridge: dw-hdmi: limit output bus formats to
 supported_formats

dw-hdmi currently builds its output bus format list from sink capabilities
and requested bpc, but it does not take bridge->supported_formats into
account. This can advertise RGB or YCbCr output formats that the encoder
side cannot accept.

Filter the returned output bus formats against supported_formats so only
formats that are supported by both the sink and the bridge are exposed.

Signed-off-by: Jonas Karlman <jonas@kwiboo.se>
---
 drivers/gpu/drm/bridge/synopsys/dw-hdmi.c | 34 +++++++++++++++--------
 1 file changed, 23 insertions(+), 11 deletions(-)

diff --git a/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c b/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c
index bfec740c6dcf..cb45c6f714af 100644
--- a/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c
+++ b/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c
@@ -2613,6 +2613,7 @@ static u32 *dw_hdmi_bridge_atomic_get_output_bus_fmts(struct drm_bridge *bridge,
 	u8 max_bpc = conn_state->max_requested_bpc;
 	bool is_hdmi2_sink = info->hdmi.scdc.supported ||
 			     (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR420));
+	unsigned int supported_formats = bridge->supported_formats;
 	u32 *output_fmts;
 	unsigned int i = 0;
 
@@ -2668,39 +2669,50 @@ static u32 *dw_hdmi_bridge_atomic_get_output_bus_fmts(struct drm_bridge *bridge,
 	 */
 
 	/* Default 8bit RGB fallback */
-	output_fmts[i++] = MEDIA_BUS_FMT_RGB888_1X24;
+	if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+		output_fmts[i++] = MEDIA_BUS_FMT_RGB888_1X24;
 
 	if (max_bpc >= 16 && info->bpc == 16) {
-		if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+		if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444) &&
+		    supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
 			output_fmts[i++] = MEDIA_BUS_FMT_YUV16_1X48;
 
-		output_fmts[i++] = MEDIA_BUS_FMT_RGB161616_1X48;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			output_fmts[i++] = MEDIA_BUS_FMT_RGB161616_1X48;
 	}
 
 	if (max_bpc >= 12 && info->bpc >= 12) {
-		if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
+		if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422) &&
+		    supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
 			output_fmts[i++] = MEDIA_BUS_FMT_UYVY12_1X24;
 
-		if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+		if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444) &&
+		    supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
 			output_fmts[i++] = MEDIA_BUS_FMT_YUV12_1X36;
 
-		output_fmts[i++] = MEDIA_BUS_FMT_RGB121212_1X36;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			output_fmts[i++] = MEDIA_BUS_FMT_RGB121212_1X36;
 	}
 
 	if (max_bpc >= 10 && info->bpc >= 10) {
-		if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
+		if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422) &&
+		    supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
 			output_fmts[i++] = MEDIA_BUS_FMT_UYVY10_1X20;
 
-		if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+		if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444) &&
+		    supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
 			output_fmts[i++] = MEDIA_BUS_FMT_YUV10_1X30;
 
-		output_fmts[i++] = MEDIA_BUS_FMT_RGB101010_1X30;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			output_fmts[i++] = MEDIA_BUS_FMT_RGB101010_1X30;
 	}
 
-	if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
+	if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422) &&
+	    supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
 		output_fmts[i++] = MEDIA_BUS_FMT_UYVY8_1X16;
 
-	if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+	if (info->color_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444) &&
+	    supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
 		output_fmts[i++] = MEDIA_BUS_FMT_YUV8_1X24;
 
 	*num_output_fmts = i;
-- 
2.43.0

